Tagging on post edit form (merging two previous PRs)

Raemon opened the pull request:
Jim and I each had made a PR attempting to address "tagging a post on the post-edit-form". We handled it in two fairly different ways. This PR basically merges the two previous PRs. (I was delighted that somehow almost all the work from the two PRs was useful and merged almost seamlessly).

I did some more UI work integrating my UI elements into Jim's architecture.

https://github.com/ForumMagnum/ForumMagnum/pull/5844
https://github.com/ForumMagnum/ForumMagnum/pull/5664

s-cheng commented on packages/lesswrong/components/form-components/FormComponentPostEditorTagging.tsx:
```diff
@@ -0,0 +1,71 @@
+}) => {
+  const { CoreTagsChecklist, TagMultiselect, FooterTagList } = Components
+  if (formType === "edit") {
+    return <FooterTagList
+      post={document}
+      hideScore
+      hidePostTypeTag
+      showCoreTags
+    />
```
I think it's important that the tags in `FooterTagList` are **not** links in this form, because clicking on it will route you to the tag page and make you lose your changes in the form.

s-cheng commented on packages/lesswrong/components/form-components/FormComponentPostEditorTagging.tsx:
```diff
@@ -0,0 +1,71 @@
+        updateCurrentValues(
+          mapValues(
+            {tagRelevance:[tag.tagId]},
+            (arrayOfTagIds: string[]) => toDictionary(
+              arrayOfTagIds, tagId=>tagId, tagId=>1
+            )
```
I feel like this should be turned into a function with a comment explaining why we're doing this

Raemon replied:
I have pinged Jim.

s-cheng replied:
I think the gist is that the post `tagRelevance` field needs to look like `{string: number}`, so even though it's extra work both here and in the callback, we need to maintain that structure.

s-cheng commented on packages/lesswrong/components/form-components/FormComponentPostEditorTagging.tsx:
```diff
@@ -0,0 +1,71 @@
+ * voting-on-tag-relevance as the post page. Styling doesn't match between these
+ * two, which is moderately unfortunate.
+ */
+const FormComponentPostEditorTagging = ({value, path, document, label, placeholder, formType, updateCurrentValues}: {
```
I'm still concerned that this is not a great mobile experience, but I assume it would be hard to hide this field on mobile and it's technically possible to downvote a tag when editing a post (so you can undo an accidental tag selection), so probably ok to leave it.

Raemon replied:
Nod. I'd prefer to err on the side of leaving mobile able to do it at-all, but we can revisit if it turns out to be annoying.

s-cheng commented on packages/lesswrong/components/tagging/FooterTagList.tsx:
```diff
@@ -68,8 +70,7 @@ const FooterTagList = ({post, classes, hideScore, hideAddTag, smallText=false}:
     limit: 100,
   });
-
-  const tagIds = (results||[]).map((tag) => tag._id)
+  const tagIds = (results||[]).map((tagRel) => tagRel.tag?._id)
```
oof was this just wrong? :(

s-cheng commented on packages/lesswrong/components/vulcan-forms/Form.tsx:
```diff
@@ -1085,6 +1085,7 @@ class Form<T extends DbObject> extends Component<any,any> {
     revertLabel: this.props.revertLabel,
     cancelCallback: this.props.cancelCallback,
     revertCallback: this.props.revertCallback,
+    submitForm: this.submitForm,
```
Do we still need this change?

Raemon replied:
It's not strictly necessary for this PR, but I do think this was a fairly obvious thing to have passed in.

Raemon replied:
It's not technically needed for this PR, but I do think it's pretty silly for the submitFormContext to not actually provide the submitForm and would prefer to leave it in.

s-cheng commented on packages/lesswrong/components/tagging/CoreTagsChecklist.tsx:
```diff
@@ -40,25 +43,20 @@ const CoreTagsChecklist = ({onSetTagsSelected, classes, post}: {
+  const handleOnTagSelected = (tag) => onTagSelected ? onTagSelected({tagId:tag._id, tagName:tag.name}) : undefined
+
+  return <>
+    {unusedCoreTags?.map(tag => <LWTooltip key={tag._id} title={<div>Click to assign <em>{tag.name}</em> tag</div>}>
```
Stray "tag", please replace with config setting

s-cheng commented on packages/lesswrong/components/tagging/CoreTagsChecklist.tsx:
```diff
@@ -17,19 +18,21 @@ const styles = (theme: ThemeType): JssStyles => ({
+  onTagSelected?: (props: {tagId: string, tagName: string})=>void,
   classes: ClassesType,
-  post: PostsList|SunshinePostsList
+  existingTagIds?: Array<string|undefined>
```
Could we make this `existingTagIds: Array<string>`?

Raemon replied:
I could fix it, but it requires either adding a filter on existingTagIds to remove any potential undefined ones, or, maybe remove the nullable option on tagRel.tagId. Nullable was added in this PR, I'm not sure why.

https://github.com/ForumMagnum/ForumMagnum/pull/3485

I don't think this is important enough to worry about for this PR, but seems good to flag it as a weird thing to think about later.

s-cheng replied:
:(

Sad, but I agree that it's not for this PR.

s-cheng reviewed (changes requested):
Looks really nice, thanks! :) The major things to address are:

1. For a new post, selecting a core tag overwrites the selected tags list
2. When editing a post, clicking on a tag routes you to the tag page, causing you to lose your changes

Raemon commented on packages/lesswrong/components/form-components/FormComponentPostEditorTagging.tsx:
```diff
@@ -0,0 +1,71 @@
+      <CoreTagsChecklist existingTagIds={Object.keys(value||{})} onTagSelected={(tag)=> {
+        updateCurrentValues(
+          mapValues(
+            {tagRelevance:[tag.tagId]},
```
Fixed.

Raemon commented on packages/lesswrong/components/posts/PostsNewForm.tsx:
```diff
@@ -12,6 +12,7 @@ import { useDialog } from "../common/withDialog";
 import { afNonMemberSuccessHandling } from "../../lib/alignment-forum/displayAFNonMemberPopups";
 import { useUpdate } from "../../lib/crud/withUpdate";
 import { useSingle } from '../../lib/crud/withSingle';
+import { groupLayoutStyles } from '../vulcan-forms/FormGroup';
```
Thanks, removed.

Raemon commented:
@s-cheng, I've addressed your comments. (Upon reflection the PR was in a pretty-obviously unfinished state, sorry for making you pay attention to a bunch of stuff I should have already caught)

s-cheng reviewed (approved):
Appreciate you addressing all the issues - it looks great and I'm excited for it to go out!
